### ISSUE TITLE
Unable to decode testnet transaction with ArrayIndexOutOfBounds Exception

### ISSUE THREAD

**Issue by Polve** [bitcoinj/bitcoinj#1348]
I'm using bitcoinj to check testnet transactions (decoding them from hex), but with this one:
https://www.blocktrail.com/tBTC/tx/95bc227eb906b4ff90039a104d78651686d2ecbc7cda602153deb72ec6938233

While doing new Transaction(...) I get an ArrayIndexOutOfBoundsException.

Is this because it's a segwit transaction?
How may I handle that problem?

**Comment by schildbach** [comment]
Can you paste your code and the stacktrace?

**Comment by Polve** [comment]
```
org.bitcoinj.core.ProtocolException: java.lang.ArrayIndexOutOfBoundsException
        at org.bitcoinj.core.Message.readBytes(Message.java:339)
        at org.bitcoinj.core.TransactionOutput.parse(TransactionOutput.java:163)
        at org.bitcoinj.core.Message.<init>(Message.java:96)
        at org.bitcoinj.core.Message.<init>(Message.java:128)
        at org.bitcoinj.core.ChildMessage.<init>(ChildMessage.java:57)
        at org.bitcoinj.core.TransactionOutput.<init>(TransactionOutput.java:79)
        at org.bitcoinj.core.Transaction.parse(Transaction.java:568)
        at org.bitcoinj.core.Message.<init>(Message.java:96)
        at org.bitcoinj.core.Message.<init>(Message.java:123)
        at org.bitcoinj.core.ChildMessage.<init>(ChildMessage.java:52)
        at org.bitcoinj.core.Transaction.<init>(Transaction.java:202)
```

**Comment by schildbach** [comment]
That's the stacktrace of the ProtocolException. What about the wrapped AIOOBException? Also, again can you post the code to reproduce this issue?

**Comment by milansismanovic** [comment]
Hi Andreas 

I faced the same issue. I am sharing my error for others to save time... :-)

Here is the code that leads to this exception:
String blockString = "00000020cb5d0...."; // whole block as a String from getblock
Block block = new Block(TestNet3Params.get(), blockString.getBytes());

Corrections convert to byte array with the correct encoding:
Block block = new Block(TestNet3Params.get(), blockString.getBytes("UTF-16"));

Please note that I am using new Block(...) which is deprecated. Will use BitcoinSerializer.makeBlock(byte[]) as soon as I find out how to create the serializer... Wanted to share the insight before this, in case I run out of time... :-)

And: Great work with the BitcoinJ library!!!!!!!!!!

Cheers,

Milan

**Comment by milansismanovic** [comment]
Hi Andreas

Actually there is still a problem:
When I get block 0000000000000d60d581450d16ae4d406d49f7049346b07e118ad480b664aeea, I get a blockString and when I deserialize it above, I get the following block.toString output:
block: 
   hash: 783f095c995122cf1f0ccb956b1a556d27504c70cab275c1200ae5ee309ef714
   version: 805371902 (BIP34, BIP66, BIP65)
   previous block: 3800360062003400360065006100610065003000320030003000300030003000
   merkle root: 3000330039003600340030006200650037003100310061003800340064003000
   time: 1711289344 (2024-03-24T14:09:04Z)
   difficulty target (nBits): 872429312
   nonce: 905984256

But it should be something like this (copied from BitcoinCore the debug window):
{
  "hash": "0000000000000d60d581450d16ae4d406d49f7049346b07e118ad480b664aeea",
  "confirmations": 569,
  "strippedsize": 567,
  "size": 567,
  "weight": 2268,
  "height": 1281625,
  "version": 536870912,
  "versionHex": "20000000",
  "merkleroot": "300e184dcee0a3c4baedb8d93d6fba4d64f7030f6914fcf0f01516fc1f5db94a",
  "tx": [
    "b56170bc4f5bf16a7d2f175c2c86a2328ba1d95480171405528647da21df343d", 
    "f1f178b39df1b58164f8167f1ae2b6fb42d0e78672f799939072d69553dce56c"
  ],
  "time": 1518415677,
  "mediantime": 1518415343,
  "nonce": 2414173497,
  "bits": "1a372b40",
  "difficulty": 304101.5177101588,
  "chainwork": "0000000000000000000000000000000000000000000000356a5215f338530813",
  "previousblockhash": "00000000000001073d717e858f94ea9181dd1c079aebe90f4c53dc6c870e5dcb",
  "nextblockhash": "00000000000004a9ee60345fae758090d5336326bbac8bdb5576bbe3b5896adb"
}

I hope you can have a look at it.

Here is my JUnit test method using ConsensusJ:
		try {
			String hashInvalidBlock = "0000000000000d60d581450d16ae4d406d49f7049346b07e118ad480b664aeea";
			String blockString = client.send("getblock", hashInvalidBlock, false);
			BitcoinSerializer serializer = new BitcoinSerializer(TestNet3Params.get(), false);
			Block block = serializer.makeBlock(blockString.getBytes(StandardCharsets.UTF_16));
			log.info(block.toString());
		} catch (Exception e) {
			e.printStackTrace();
		}

I have checked the blockString byte code and it is identical like the one in the BitcoinCore debug window.

**Comment by milansismanovic** [comment]
I can now narrow it down to transaction deserialization:
Creating a BitcoinJ Transaction from a byte array fails for some Transaction I got through JSON/RPC calling its getrawtransaction service from a testnet instance.

The code looked something like this:
for (String transactionHash : transactions) {
	String rawTransaction = client.getrawtransaction(transactionHash);
	byte[] rawTransactionBytes = Utils.HEX.decode(rawTransaction);
	try {
		Transaction bitcoinjTx = new Transaction(TestNet3Params.get(), rawTransactionBytes);
		log.info(bitcoinjTx.toString());
	} catch (Exception e) {
		log.info("failed tx deser: {}, bytes: {}", e.getMessage(), rawTransaction);
	}
...
}

Here is the log from the entries that failed attached to this message.
[transaction-deser-problem.txt](https://github.com/bitcoinj/bitcoinj/files/1726607/transaction-deser-problem.txt)

I can decode them in the Bitcon Core debug console using decoderawtransaction. Am I missing something?

Thank you for having a look.

Cheers,

Milan

**Comment by milansismanovic** [comment]
OK, I think I narrowed it down to segwit RawTransactions, that cannot be deserialized. I come to this conclusion, because all the ones that had an error contained a txinwitness when deserialized in the Bitcoin Core Debug console.

I have added the raw transactions with unique error messages into the attached JUnit test.

**Could someone please confirm: BitcoinJ doesn't support deserialization of transactions, that have segwit in it.**

[BitcoinTransactionDeserialization.zip](https://github.com/bitcoinj/bitcoinj/files/1737789/BitcoinTransactionDeserialization.zip)

**Comment by schildbach** [comment]
@milansismanovic Confirmed. At least on any released version and current master. You can try the segwit branch though, although it's likely unstable.

**Comment by milansismanovic** [comment]
Thank you for confirming. I started to doubt my sanity... :-)

Will eventually investigate the segwit branch.

**Comment by msgilligan** [comment]
If you're accessing the blocks via `bitcoind` over JSON-RPC, you could try the `-rpcserialversion=0` command-line option as a workaround. We used this in **OmniJ**:
https://github.com/OmniLayer/OmniJ/commit/8946dc38ef58438f0fda17b21b57deeb2908e9c0

**Comment by vaustymenko** [comment]
I ran into the same issue. Are there any workarounds?

**Comment by schildbach** [comment]
This should actually be fixed on master, thanks to commit 648655da99ac7cdc48fef7d5b6aaa57148af3d6f. I'll close this. Please test and feel free to comment again if there is still a problem.

**Comment by barrongineer** [comment]
Am I the only one still having issues with this? I'm on master branch and still receiving the error.

Some kotlin in a unit test:
```kotlin
val params = RegTestParams.get()
Context.getOrCreate(params)

val to = Address.fromString(params, "mxYDYiWDXFob4VwmYzLgJz96iYjXGsZphY")
val sendRequest = SendRequest.to(to, Coin.valueOf(1000000))
val serialized = sendRequest.tx.bitcoinSerialize()
val tx = Transaction(params, serialized)
```

Stacktrace produced from the `Transaction` constructor:
```
org.bitcoinj.core.ProtocolException: java.lang.ArrayIndexOutOfBoundsException: 44

	at org.bitcoinj.core.Message.readUint32(Message.java:298)
	at org.bitcoinj.core.TransactionInput.parse(TransactionInput.java:157)
	at org.bitcoinj.core.Message.<init>(Message.java:97)
	at org.bitcoinj.core.Message.<init>(Message.java:129)
	at org.bitcoinj.core.ChildMessage.<init>(ChildMessage.java:57)
	at org.bitcoinj.core.TransactionInput.<init>(TransactionInput.java:146)
	at org.bitcoinj.core.Transaction.parseInputs(Transaction.java:611)
	at org.bitcoinj.core.Transaction.parse(Transaction.java:593)
	at org.bitcoinj.core.Message.<init>(Message.java:97)
	at org.bitcoinj.core.Message.<init>(Message.java:124)
	at org.bitcoinj.core.ChildMessage.<init>(ChildMessage.java:52)
	at org.bitcoinj.core.Transaction.<init>(Transaction.java:218)
	at c.c.coldwallet.service.BitcoinWalletServiceTest.testTransactionConstructor(BitcoinWalletServiceTest.kt:76)
```
